### PR TITLE
Sdp overhaul

### PR DESCRIPTION
**Pull request by bbaldino**
can ignore this for now.  want to run the smoke tests...some are failing on my instance but i'm wondering if it's a config thing.

### CODE REVIEW

**Comment by bbaldino** [comment]
cool...guess there must be a config difference on my jitsi meet dev server that causes smoke tests to fail.  @hristoterezov do you want to take a look at this?  there's still more cleanup to do, so no need to merge yet.  but the idea is this:

- rework the modify sources queue / processing loop: rather than having logic in the processing loop, move the logic to the tasks themselves and keep the loop simple. 
- add a shim to anything still using the modifySources queue to keep things working as expected
- change the 'video camera device change' flow to use the new loop directly.  the other scenarios (video mute, toggle screenshare, etc.) will eventually all be moved over to this style.
- because of the shim, both old and new scenarios are processed through the same queue, so shouldn't be any out-of-order or concurrency problems

this is a small piece of a larger project to get some things cleaned up and more efficient (no intermediary states when doing compound operations like toggle screenshare or change camera), so expect even the code in this diff to evolve in future PRs as i get to more of the scenarios.

**Comment by bbaldino** [comment]
@hristoterezov this is officially ready for you to take a look at :) we can go through it in person tomorrow and i can give you a summary.  tests are passing and i've bounced some ideas off of george, so should be in pretty decent shape already.

File: modules/xmpp/SdpConsistency.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
+var transform = require('sdp-transform');
```
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: Why not using `import`? 

Also maybe transform should be declared with `const` or `let` if we are using ES6.

File: modules/xmpp/SdpConsistency.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+/**
+ * Begin helper functions
+ */
+function getPrimarySsrc (videoMLine) {
```
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: JS docs missing

File: modules/xmpp/SdpConsistency.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            }
+            let fidGroup = findGroup(videoMLine, "FID");
+            if (fidGroup) {
+                return parseInt(fidGroup.ssrcs.split(" ")[0]);
+            }
+        }
+    }
+}
+
+function getRtxSsrc (videoMLine, primarySsrc) {
```
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: JS docs missing

File: modules/xmpp/SdpConsistency.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                        }
+                    });
+                }
+            }
+        }
+        return transform.write(parsedSdp);
+    }
+}
+
+module.exports = SdpConsistency;
```
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: Why not `export default`?

File: modules/xmpp/SdpConsistency.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+/**
+ * Handles the work of keeping video ssrcs consistent across multiple
+ * o/a cycles, making it such that all stream operations can be
+ * kept local and do not need to be signaled.
+ * NOTE: This only keeps the 'primary' video ssrcs consistent: meaning
+ * the primary video stream and an associated RTX stream, if it exists
+ */
+class SdpConsistency {
+    constructor () {
```
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: JS docs missing

File: modules/xmpp/SdpConsistency.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        this.clearSsrcCache();
+    }
+
+    clearSsrcCache () {
```
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: JS docs missing

File: modules/xmpp/SdpConsistency.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        this.cachedPrimarySsrc = null;
+        this.cachedPrimaryRtxSsrc = null;
+    }
+
+    setPrimarySsrc (primarySsrc) {
```
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: JS docs missing

File: modules/xmpp/SdpConsistency.js [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        this.cachedPrimarySsrc = primarySsrc;
+    }
+
+    makeVideoPrimarySsrcsConsistent (sdpStr) {
```
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: JS docs missing

File: JitsiConference.js [hunk 9]
```diff
@@ -552,24 +500,98 @@ JitsiConference.prototype.removeTrack = function (track) {
...
+        .then(() => {
+            if (newStream) {
+                // Now handle the addition of the newStream at the JitsiConference level
+                this._setupNewTrack(newStream);
+            }
+            return Promise.resolve();
+        });
+};
+
+JitsiConference.prototype._setupNewTrack = function (newTrack) {
```
Comment (hristoterezov) [inline JitsiConference.js]: JS docs missing

File: modules/xmpp/JingleSessionPC.js [hunk 10]
```diff
@@ -600,34 +552,9 @@ JingleSessionPC.prototype.onTerminated = function (reasonCondition,
...
-    logger.log('addssrc', new Date().getTime());
-    logger.log('ice', this.peerconnection.iceConnectionState);
-
-    this.readSsrcInfo(elem);
-
-    var sdp = new SDP(this.peerconnection.remoteDescription.sdp);
-    var mySdp = new SDP(this.peerconnection.localDescription.sdp);
-
-    $(elem).each(function (idx, content) {
+JingleSessionPC.prototype._parseSsrcInfoFromSourceAdd = function (sourceAddElem, currentRemoteSdp) {
```
Comment (hristoterezov) [inline modules/xmpp/JingleSessionPC.js]: JS docs missing

File: modules/xmpp/JingleSessionPC.js [hunk 11]
```diff
@@ -663,51 +581,256 @@ JingleSessionPC.prototype.addSource = function (elem) {
...
+ * });
+ */
+JingleSessionPC.prototype._processQueueTasks = function (task, finishedCallback) {
+    task(finishedCallback);
+};
+
+/**
+ * Takes in a jingle offer iq, returns the new sdp offer
+ */
+JingleSessionPC.prototype._processNewJingleOfferIq = function(offerIq) {
```
Comment (hristoterezov) [inline modules/xmpp/JingleSessionPC.js]: `@param` and `@returns` docs missing

File: modules/xmpp/JingleSessionPC.js [hunk 12]
```diff
@@ -663,51 +581,256 @@ JingleSessionPC.prototype.addSource = function (elem) {
...
+    if (this.failICE) {
+        remoteSdp.failICE = true;
+    }
+
+    remoteSdp.fromJingle(offerIq);
+    this.readSsrcInfo($(offerIq).find(">content"));
+    return remoteSdp;
+};
+
+JingleSessionPC.prototype._processRemoteRemoveSource = function (removeSsrcInfo) {
```
Comment (hristoterezov) [inline modules/xmpp/JingleSessionPC.js]: JS docs missing

File: modules/xmpp/JingleSessionPC.js [hunk 13]
```diff
@@ -663,51 +581,256 @@ JingleSessionPC.prototype.addSource = function (elem) {
...
+            remoteSdp.media[idx] = remoteSdp.media[idx].replace(line + '\r\n', '');
+        });
+    });
+    remoteSdp.raw = remoteSdp.session + remoteSdp.media.join('');
+
+    return remoteSdp;
+};
 
-    $(elem).each(function (idx, content) {
+JingleSessionPC.prototype._processRemoteAddSource = function (addSsrcInfo) {
```
Comment (hristoterezov) [inline modules/xmpp/JingleSessionPC.js]: JS docs missing

File: modules/xmpp/JingleSessionPC.js [hunk 14]
```diff
@@ -663,51 +581,256 @@ JingleSessionPC.prototype.addSource = function (elem) {
...
+        remoteSdp.media[idx] += lines;
+    });
+    remoteSdp.raw = remoteSdp.session + remoteSdp.media.join('');
+
+    return remoteSdp;
+};
+
+/**
+ * Do a new o/a flow using the existing remote description
+ * @param sdp: optional remote sdp to use.  if not provided, the remote
```
Comment (hristoterezov) [inline modules/xmpp/JingleSessionPC.js]: It's good to have the exact name of the parameter and type:
`@param {Type} exampleVar <description>`

Also `@returns` is missing.

File: modules/xmpp/JingleSessionPC.js [hunk 15]
```diff
@@ -663,51 +581,256 @@ JingleSessionPC.prototype.addSource = function (elem) {
...
+ *  by the modification queue.
+ */
+JingleSessionPC.prototype.addStreamToPeerConnection = function (stream, ssrcInfo) {
+    let actualStream = stream && stream.getOriginalStream ? stream.getOriginalStream() : stream;
+    if (this.peerconnection) {
+        this.peerconnection.addStream(actualStream, ssrcInfo);
+    }
+};
+
+JingleSessionPC.prototype._parseSsrcInfoFromSourceRemove = function (sourceRemoveElem, currentRemoteSdp) {
```
Comment (hristoterezov) [inline modules/xmpp/JingleSessionPC.js]: JS docs missing

File: modules/xmpp/SdpConsistency.js (hunk 1 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: done (changed to use import)

File: modules/xmpp/SdpConsistency.js (hunk 2 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: done

File: modules/xmpp/SdpConsistency.js (hunk 3 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: done

File: modules/xmpp/SdpConsistency.js (hunk 5 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: i already have a js doc above the class, is there something specific in there that i'm missing?

File: modules/xmpp/SdpConsistency.js (hunk 6 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: done

File: modules/xmpp/SdpConsistency.js (hunk 7 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: done

File: modules/xmpp/SdpConsistency.js (hunk 8 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: done

File: modules/xmpp/SdpConsistency.js (hunk 4 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: done

File: JitsiConference.js (hunk 9 above)
Comment (bbaldino) [inline JitsiConference.js]: done

File: modules/xmpp/JingleSessionPC.js (hunk 10 above)
Comment (bbaldino) [inline modules/xmpp/JingleSessionPC.js]: done

File: modules/xmpp/JingleSessionPC.js (hunk 11 above)
Comment (bbaldino) [inline modules/xmpp/JingleSessionPC.js]: done

File: modules/xmpp/JingleSessionPC.js (hunk 12 above)
Comment (bbaldino) [inline modules/xmpp/JingleSessionPC.js]: done

File: modules/xmpp/JingleSessionPC.js (hunk 13 above)
Comment (bbaldino) [inline modules/xmpp/JingleSessionPC.js]: done

File: modules/xmpp/JingleSessionPC.js (hunk 14 above)
Comment (bbaldino) [inline modules/xmpp/JingleSessionPC.js]: done

File: modules/xmpp/JingleSessionPC.js (hunk 15 above)
Comment (bbaldino) [inline modules/xmpp/JingleSessionPC.js]: done

File: JitsiConference.js [hunk 16]
```diff
@@ -552,24 +500,98 @@ JitsiConference.prototype.removeTrack = function (track) {
...
-    if(!this.room){
-        if(this.rtc) {
-            this.onTrackRemoved(track);
+/**
+ * Replaces oldStream with newStream and performs a single offer/answer
+ *  cycle after both operations are done.  Either oldStream or newStream
+ *  can be null; replacing a valid 'oldStream' with a null 'newStream'
+ *  effectively just removes 'oldStream'
+ * @param oldStream the current stream in use to be replaced
+ * @param newStream the new stream to use
```
Comment (bbaldino) [inline JitsiConference.js]: done

File: modules/xmpp/JingleSessionPC.js [hunk 17]
```diff
@@ -322,73 +327,20 @@ JingleSessionPC.prototype.acceptOffer = function(jingleOffer,
...
+                success();
+                finishedCallback();
+            }, (error) => {
+                logger.info("Error renegotiating after setting new remote offer: " + error);
+                failure(error);
+                JingleSessionPC.onJingleFatalError(this, error);
+                finishedCallback(error);
+            });
+    };
+    this.modificationQueue.push(workFunction);
```
Comment (bbaldino) [inline modules/xmpp/JingleSessionPC.js]: yeah i just thought this way looked cleaner and the name makes it more obvious what's being done i thought.  i can change it if you don't like it this way though.

File: modules/xmpp/JingleSessionPC.js [hunk 18]
```diff
@@ -663,51 +581,256 @@ JingleSessionPC.prototype.addSource = function (elem) {
...
+ *  cycle after both operations are done.  Either oldStream or newStream
+ *  can be null; replacing a valid 'oldStream' with a null 'newStream'
+ *  effectively just removes 'oldStream'
+ * @param oldStream the current stream in use to be replaced
+ * @param newStream the new stream to use
+ * @returns {Promise}
+ */
+JingleSessionPC.prototype.replaceStream = function (oldStream, newStream) {
+    return new Promise((resolve, reject) => {
+        let workFunction = (finishedCallback) => {
```
Comment (bbaldino) [inline modules/xmpp/JingleSessionPC.js]: same as my comment above...i just think it's easier to read and is more self-documenting with the 'workFunction' name.  i can change it though.

File: JitsiConference.js [hunk 19]
```diff
@@ -425,16 +426,6 @@ JitsiConference.prototype.addTrack = function (track) {
...
-        if(ssrcMap[this.getMSID()]){
-            this._setSSRC(ssrcMap[this.getMSID()]);
-            conference.room.removeListener(XMPPEvents.SENDRECV_STREAMS_CHANGED,
-                this.ssrcHandler);
-        }
-    }.bind(track, this);
-    this.room.addListener(XMPPEvents.SENDRECV_STREAMS_CHANGED,
-        track.ssrcHandler);
-
     if(track.isAudioTrack() || (track.isVideoTrack() &&
```
Comment (bbaldino) [inline JitsiConference.js]: yeah this is a bit awkward i guess...we don't call this method anywhere anymore but it was kept to not break the API.  Basically it gets called here as well because you don't call both flow (addTrack and replaceStream) for the same track...it's one or the other.

Comment (bbaldino) [inline JitsiConference.js]: oops...forgot we called replaceStream from there.  ignore that last comment :)

Comment (bbaldino) [inline JitsiConference.js]: done

File: JitsiConference.js [hunk 20]
```diff
@@ -552,24 +500,98 @@ JitsiConference.prototype.removeTrack = function (track) {
...
+/**
+ * Replaces oldStream with newStream and performs a single offer/answer
+ *  cycle after both operations are done.  Either oldStream or newStream
+ *  can be null; replacing a valid 'oldStream' with a null 'newStream'
+ *  effectively just removes 'oldStream'
+ * @param oldStream the current stream in use to be replaced
+ * @param newStream the new stream to use
+ * @returns {Promise}
+ */
+JitsiConference.prototype.replaceStream = function (oldStream, newStream) {
```
Comment (bbaldino) [inline JitsiConference.js]: there is quite a bit of inconsistency in the usage of track vs. stream.  i started with 'replaceTrack' but all the methods beneath it that are called are named with 'stream'.

fixed the disposed checks

File: modules/xmpp/JingleSessionPC.js [hunk 21]
```diff
@@ -663,51 +581,256 @@ JingleSessionPC.prototype.addSource = function (elem) {
...
+                        let localUfrag = getUfrag(answer.sdp);
+                        if (localUfrag != this.localUfrag) {
+                            this.localUfrag = localUfrag;
+                            this.room.eventEmitter.emit(
+                                    XMPPEvents.LOCAL_UFRAG_CHANGED, localUfrag);
+                        }
+                        this.peerconnection.setLocalDescription(
+                            answer,
+                            () => { resolve(); },
+                            (error) => { reject(error, "setLocalDescription failed"); }
```
Comment (bbaldino) [inline modules/xmpp/JingleSessionPC.js]: done

File: modules/xmpp/JingleSessionPC.js [hunk 22]
```diff
@@ -663,51 +581,256 @@ JingleSessionPC.prototype.addSource = function (elem) {
...
+                        );
+                    },
+                    (error) => { reject(error, "createAnswer failed"); },
```
Comment (bbaldino) [inline modules/xmpp/JingleSessionPC.js]: done

File: modules/xmpp/JingleSessionPC.js [hunk 23]
```diff
@@ -663,51 +581,256 @@ JingleSessionPC.prototype.addSource = function (elem) {
...
+                    media_constraints
+                );
+            },
+            (error) => { reject(error, "setRemoteDescription failed"); }
```
Comment (bbaldino) [inline modules/xmpp/JingleSessionPC.js]: done

File: JitsiConference.js [hunk 24]
```diff
@@ -552,24 +500,98 @@ JitsiConference.prototype.removeTrack = function (track) {
...
+    // First do the removal of the oldStream at the JitsiConference level
+    if (oldStream) {
+        this.onTrackRemoved(oldStream);
```
Comment (bbaldino) [inline JitsiConference.js]: done

File: modules/xmpp/SdpConsistency.js (hunk 5 above)
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: We are usually adding JS doc for the contructor as well.

File: JitsiConference.js (hunk 20 above)
Comment (hristoterezov) [inline JitsiConference.js]: At list for all public APIs we have only the concept of tracks I think. And the objects that you are passing are of type JitsiLocalTrack so ...

File: modules/xmpp/JingleSessionPC.js (hunk 17 above)
Comment (hristoterezov) [inline modules/xmpp/JingleSessionPC.js]: I prefer the other way personally.

File: modules/xmpp/JingleSessionPC.js (hunk 18 above)
Comment (hristoterezov) [inline modules/xmpp/JingleSessionPC.js]: I prefer the other way personally.

File: modules/xmpp/SdpConsistency.js [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+ *  (if there is one) for that video ssrc
+ * @param {object} videoMLine the video MLine from which to extract the
+ *  rtx video ssrc
+ * @param {number} primarySsrc the video ssrc for which to find the
+ *  corresponding rtx ssrc
+ * @returns {number} the rtx ssrc (or undefined if there isn't one)
+ */
+function getRtxSsrc (videoMLine, primarySsrc) {
+    if (videoMLine.ssrcGroups) {
+        return videoMLine
```
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: Personally I would implement this in the following way:
```
let res;
videoMLine.ssrcGroups.find(ssrcGroup => {
    if(ssrcGroup.semantics === "FID"
        &&  parseInt(ssrcGroup.ssrcs.split(" ")[0]) === primarySsrc) {
        res = parseInt(ssrcGroup.ssrcs.split(" ")[1]);
        return true; 
    }
    return false;
})
```
or with simple loop or for Each.
This way it should do less operations (in this case maybe it won't matter) and also personally I find it easier to understand.

File: modules/xmpp/SdpConsistency.js [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+ * @returns {number} the primary video ssrc
+ */
+function getPrimarySsrc (videoMLine) {
+    if (!videoMLine.ssrcs) {
+        return;
+    }
+    let numSsrcs = videoMLine.ssrcs
+        .filter(ssrcInfo => ssrcInfo.attribute === "msid")
+        .map(ssrcInfo => ssrcInfo.id)
+        .filter((ssrc, index, array) => array.indexOf(ssrc) === index)
```
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: Why do we need the last filter?

File: modules/xmpp/SdpConsistency.js [hunk 27]
```diff
@@ -0,0 +1,176 @@
...
+        .length;
+    if (numSsrcs === 1) {
+        return videoMLine.ssrcs
```
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: Isn't it:
```
return videoMLine.ssrcs
           .filter(ssrcInfo => ssrcInfo.attribute === "msid")[0].id
```
simpler.

Also If we have more than one places in the function where we use `videoMLine.ssrcs.filter(ssrcInfo => ssrcInfo.attribute === "msid")` it's better to execute it once and assign it to a variable.

File: modules/xmpp/SdpConsistency.js (hunk 25 above)
Comment (hristoterezov) [inline modules/xmpp/SdpConsistency.js]: or:
```
let group = videoMLine.ssrcGroups.find(ssrcGroup => {
    if(ssrcGroup.semantics === "FID"
        &&  parseInt(ssrcGroup.ssrcs.split(" ")[0]) === primarySsrc) {
        return true; 
    }
    return false;
});
return  group? parseInt(group.ssrcs.split(" ")[1]) : undefined;
```

Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: i don't find the above to be very easy to read (primarily because of the long line with the call to parseInt inside the if), and i think assigning to an outside variable inside the 'find' doesn't feel great, but i did change this to use find to find the specific group and then move the rtxSsrc parsing (once the group has been found) outside so the .map/.filter chain is gone.

File: modules/xmpp/SdpConsistency.js (hunk 26 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: this list contains multiple entries for each ssrc (because it's each ssrc line in the sdp, so has one for msid, one for cname, etc.) so the last line ensures only unique entries are added.

Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: ah....oops...yes i see i am double filtering here...i think i forgot to remove the msid filtering line before (i think it's cleaner not to assume there is an msid line necessarily).  i'll clean this up.

File: modules/xmpp/SdpConsistency.js (hunk 27 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: i think we can make this even simpler actually, if there is only 1 ssrc we can just do:

  return videoMLine.ssrcs[0].id;

**Comment by paweldomas** [comment]
@bbaldino I just went through the whole PR and I find it very interesting (and nice) how you have separated the changes into commits by providing intermediate shims for the code that is not complete yet. Also I like the modify sources cleanup that it submits actual task functions rather than executing all the stuff in modify sources method. The replaceStreams method is also something I've been thinking about for a while never getting to it. All these changes are big improvement to the current code. I have few question about some changes, but will post them on per commit basis.

File: modules/xmpp/SdpConsistency.js [hunk 28]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    /**
+     * Explicitly set the primary ssrc to be used in
+     *  makeVideoPrimarySsrcsConsistent
+     * @param {number} primarySsrc the primarySsrc to be used
+     *  in future calls to makeVideoPrimarySsrcsConsistent
+     */
+    setPrimarySsrc (primarySsrc) {
+        this.cachedPrimarySsrc = primarySsrc;
```
Comment (paweldomas) [inline modules/xmpp/SdpConsistency.js]: I haven't looked at the simulcast lib if there are any changes related to SSRC caching, but I wonder if the non-primary SSRCs will be consistent as well ? Does the simulcast lib cache non-primary SSRCs as long as the primary one remains the same ?

Comment (paweldomas) [inline modules/xmpp/SdpConsistency.js]: Oh actually I see that now the SSRC cache is being set on the simulcast lib when doing "unmute"

File: modules/xmpp/JingleSessionPC.js (hunk 18 above)
Comment (paweldomas) [inline modules/xmpp/JingleSessionPC.js]: I like the approach with var named worked function. As Brian wrote it self documents the fact that this is a worker function that is about to be submitted to the modification queue.

File: modules/xmpp/SdpConsistency.js (hunk 28 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: yup...the division of responsibilities is:
SdpConsistency keep the primary video ssrc consistent
SdpSimulcast adds simulcast ssrcs and takes the responsibility for keeping the sim ssrcs consistent
(And, in a future PR: RtxModifier adds rtx ssrcs and keeps them consistent)

File: modules/xmpp/JingleSessionPC.js [hunk 29]
```diff
@@ -663,51 +591,285 @@ JingleSessionPC.prototype.addSource = function (elem) {
...
+                    finishedCallback();
+                });
+        };
+        this.modificationQueue.push(workFunction);
+    });
+};
+
+/**
+ * Just add the stream to the peerconnection
+ * @param stream either the low-level webrtc MediaStream or
```
Comment (paweldomas) [inline modules/xmpp/JingleSessionPC.js]: a correct way for documenting this would be:
`@param {JitsiLocalTrack|MediaStream} stream either the low-level webrtc MediaStream or...`
this way some IDEs will recognise the type and provide code completion.

File: modules/xmpp/SdpConsistency.js [hunk 30]
```diff
@@ -55,13 +52,15 @@ function getPrimarySsrc (videoMLine) {
...
 function getRtxSsrc (videoMLine, primarySsrc) {
     if (videoMLine.ssrcGroups) {
-        return videoMLine
-            .ssrcGroups
-            .filter(ssrcGroup => ssrcGroup.semantics === "FID")
-            .filter(ssrcGroup => 
-                parseInt(ssrcGroup.ssrcs.split(" ")[0]) === primarySsrc)
-            .map(ssrcGroup => ssrcGroup.ssrcs.split(" ")[1])
-            .map(ssrcStr => parseInt(ssrcStr))[0];
+                let fidGroup = videoMLine.ssrcGroups.find(group => {
```
Comment (paweldomas) [inline modules/xmpp/SdpConsistency.js]: Not a big deal, but the indentation seems to be messed up.

File: modules/xmpp/JingleSessionPC.js (hunk 29 above)
Comment (bbaldino) [inline modules/xmpp/JingleSessionPC.js]: fixed

File: modules/xmpp/SdpConsistency.js (hunk 30 above)
Comment (bbaldino) [inline modules/xmpp/SdpConsistency.js]: oops...thanks.  fixed.

File: modules/xmpp/JingleSessionPC.js [hunk 31]
```diff
@@ -962,98 +963,106 @@ JingleSessionPC.prototype.generateNewStreamSSRCInfo = function () {
...
+ * NOTE: must be called within a work function being executed
+ *  by the modification queue.
+ */
+JingleSessionPC.prototype.removeStreamFromPeerConnection = function (stream, stopStream) {
+    let actualStream = stream && stream.getOriginalStream ? stream.getOriginalStream() : stream;
+    if (!this.peerconnection) {
         return;
     }
-
     if (RTCBrowserType.getBrowserType() ===
```
Comment (paweldomas) [inline modules/xmpp/JingleSessionPC.js]: I know that it was like that before the changes and I'm not asking you to change that now, but in general it seems to make more sense to distinguish such cases on per feature basis rather than on per browser type basis. In this case the RTP sender API is used to avoid sLD/sRD cycle. Hence it would make sense to add a method to BrowserType.isRTPSenderAPISupported() and include check for Firefox browser type there. This way it will be easier to change than in future and keep the logic in one place. I did something like that for the onmute event support when a video track is frozen: https://github.com/jitsi/lib-jitsi-meet/blob/master/modules/RTC/RTCBrowserType.js#L125

File: modules/xmpp/JingleSessionPC.js [hunk 32]
```diff
@@ -724,173 +886,24 @@ JingleSessionPC.prototype.removeSource = function (elem) {
...
-
-    /**
-     * Implements a failure callback which reports an error message and an
-     * optional error through (1) logger, (2) GlobalOnErrorHandler, and (3)
-     * queueCallback.
-     *
-     * @param {string} errmsg the error message to report
-     * @param {*} error an optional error to report in addition to errmsg
-     */
-    function reportError(errmsg, err) {
```
Comment (hristoterezov) [inline modules/xmpp/JingleSessionPC.js]: Why not using this code on some of the errors from SLD, SRD, etc?

File: modules/xmpp/JingleSessionPC.js (hunk 31 above)
Comment (bbaldino) [inline modules/xmpp/JingleSessionPC.js]: yeah i think that sounds like a better take on it, if you don't think it's too bad for now though i agree it'd be nice to have that in a separate PR

File: JitsiConference.js (hunk 20 above)
Comment (bbaldino) [inline JitsiConference.js]: fixed this (change to replaceTrack)
